UMAprotocol/snapshot
# feat: import transactions from Safe builder

Fixes #UMA-2431

## motivation
We want users to be able to use the transaction builder in the safe app, export it, and then import it into the oSnap plugin.

## How to test
1. create a new Proposal (check "use oSnap" checkbox)
2. Add a new transaction
3. Choose transaction type "Safe Import" from list
4. either drag n drop or select file from machine. (see 2 test files below)
5. there are multiple transactions in the file. choose 1 to import.
6. There are 2 transaction types **Native transfer** & **contract interaction**
7. The relevant input fileds should be populated with the values from the file and validated (**it is possible to export invalid values from safe's builder**, so we must validate each input here in the app)

## Test files
This file works ✅ 
[test-file-VALID.json](https://github.com/UMAprotocol/snapshot/files/14667816/test-file-VALID.json)

This file is corrupted ❌ (should display an error message)
[test-file-INVALID.json](https://github.com/UMAprotocol/snapshot/files/14667808/test-file-INVALID.json)

screenshots:
> idle state
![Screenshot 2024-03-20 at 13 43 15](https://github.com/UMAprotocol/snapshot/assets/51655063/0cac8d06-67f9-4ea6-9361-6e04a80fee23)
> add file, choose transaction
![Screenshot 2024-03-20 at 13 43 31](https://github.com/UMAprotocol/snapshot/assets/51655063/93bc8d80-aa9d-4842-a538-6fb91c49e4d4)
> Native transfer
![Screenshot 2024-03-20 at 13 43 58](https://github.com/UMAprotocol/snapshot/assets/51655063/20cf6540-6ec6-4104-bcc1-d80495e29384)
> Contract Interaction (includes erc20 transfers)
![Screenshot 2024-03-20 at 13 43 51](https://github.com/UMAprotocol/snapshot/assets/51655063/245a7491-cf3e-46e1-b545-baaaf106e13d)
> If you drag n drop a non-json file
![Screenshot 2024-03-20 at 14 12 25](https://github.com/UMAprotocol/snapshot/assets/51655063/15e8ff3d-ddd7-4311-a18a-464682aa5664)
 > If there is an error while attempting to parse the file. can be syntax errors in the file or the schema does not match what we expect.
![Screenshot 2024-03-20 at 14 13 19](https://github.com/UMAprotocol/snapshot/assets/51655063/de25f411-f4ab-4fbc-9fd9-295fac9342e5)
